Kafka Connect: Commit coordination

bryanck opened the pull request:
This PR is the next stage in submitting the Kafka Connect Iceberg sink connector, and is a follow up to https://github.com/apache/iceberg/pull/8701, https://github.com/apache/iceberg/pull/9466, and https://github.com/apache/iceberg/pull/9641. It includes the commit coordinator and related tests.

Still not included for the sink are the integration tests, distribution build, or docs, which will be added in follow up PRs. For reference, the current sink implementation can be found at https://github.com/tabular-io/iceberg-kafka-connect, and you can read some existing docs at https://github.com/tabular-io/iceberg-kafka-connect/tree/main/docs.

ajantha-bhat commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:
```diff
@@ -0,0 +1,294 @@
+            .filter(distinctByKey(dataFile -> dataFile.path().toString()))
+            .collect(Collectors.toList());
+
+    List<DeleteFile> deleteFiles =
```
Since we are only supporting append, do we need this code?

ajantha-bhat replied:
Similar comments for below `RowDelta` block.

bryanck replied:
I wanted to keep the coordinator capable of handling delete files, so when we do add in delta support, it should be fairly straightforward, and won't require changes to the control message data model.

ajantha-bhat commented on kafka-connect/kafka-connect/src/test/java/org/apache/iceberg/connect/channel/CoordinatorTest.java:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  @Test
+  public void testCommitDelta() {
```
maybe delete file tests can be added when we have delete writers feature added.

bryanck replied:
See my comment above, I'd prefer to add this into the coordinator now, rather than later.

ajantha-bhat replied:
Ack

fqaiser94 commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+
+  @Override
+  public void open(Collection<TopicPartition> partitions) {
+    catalog = CatalogUtils.loadCatalog(config);
```
`open` method is called with only the newly assigned partitions. Is there a strong reason to pass just the _newly assigned_ partitions to the `Committer.start` method when the Committer can just retrieve _all_ partitions assigned to this task via `context.assignment` anyway?

I'm also worried we might have a bug here. The Committer implementation uses this `partitions` argument to check if partition 0 of the first topic is assigned to this task and if so, it spawns a Coordinator process. I'm worried that if there was a rebalance where the partition 0 of the first topic doesn't move between tasks, then it would not be included in the `partitions` argument for any Task and thus we could potentially end up with a Connector that doesn't have any Coordinator process running on any Task. Thoughts?

bryanck replied:
That's not my understanding, `open()` will be called with the new assignment, i.e. all assigned topic partitions. See the [javadoc](https://kafka.apache.org/37/javadoc/org/apache/kafka/connect/sink/SinkTask.html#open(java.util.Collection)): "The list of partitions that are now assigned to the task (may include partitions previously assigned to the task)"

fqaiser94 replied:
I did read the javadoc and interpreted it differently but I can totally see how you reached your interpretation as well. 

The reason I'm inclined towards my interpretation is because I know `SinkTask.open` is ultimately called by the `ConsumerRebalanceListener.onPartitionsAssigned` method ([here](https://github.com/apache/kafka/blob/958bc0601c7852b8aaf28960280b1f45efd7d850/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L741) in the Kafka source code) and if you look at the javadocs for that method, it says `The list of partitions that are now assigned to the consumer (previously owned partitions will NOT be included, i.e. this list will only include newly added partitions)`. 

I could still be wrong so feel free to test things out or ask in the kafka community.

bryanck replied:
Thanks, it does seem like you're right. I changed this to get the consumer assignment instead.

fqaiser94 commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:
```diff
@@ -0,0 +1,140 @@
+        new TopicPartition(record.topic(), record.kafkaPartition()),
+        new Offset(record.kafkaOffset() + 1, timestamp));
+
+    if (config.dynamicTablesEnabled()) {
```
Same here, we really don't need to check this on every record?

Feels like we're missing an abstraction here, something like the concept of a `Router` which has a `StaticRouter` implementation and a `DynamicRouter` implementation and only one of those is constructed for the lifetime of a `SinkWriter` based on the `config.dynamicTablesEnabled()` setting.

ajantha-bhat replied:
Agree. But IMO we should avoid too much internal refactoring which makes extracting code from the Tabular implementation harder. 

We can always refactor once the complete code is in.

bryanck replied:
I think this is a good suggestion, and I looked into adding an abstraction here. Ultimately I felt it added complexity, so I agree we should leave it as-is for now.

fqaiser94 commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:
```diff
@@ -0,0 +1,140 @@
+    Preconditions.checkNotNull(routeField, "Route field cannot be null with dynamic routing");
+
+    String routeValue = extractRouteValue(record.value(), routeField);
+    if (routeValue != null) {
```
IMO we should throw an error instead of dropping the record.
Users can filter out messages easily using an SMT to get the same behaviour, if necessary.
In the future, I imagine we can allow users to supply custom exception-handler implementations which could also allow users to drop records on error.

bryanck replied:
I feel skipping is better than getting into a state where the sink can no longer progress. When we add DLQ support then we could route to that.

ajantha-bhat replied:
+1, we can add a TODO about DLQ here or an issue for tracking it.

fqaiser94 commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:
```diff
@@ -0,0 +1,140 @@
+    String routeField = config.tablesRouteField();
+    Preconditions.checkNotNull(routeField, "Route field cannot be null with dynamic routing");
+
+    String routeValue = extractRouteValue(record.value(), routeField);
+    if (routeValue != null) {
+      String tableName = routeValue.toLowerCase();
+      writerForTable(tableName, record, true).write(record);
+    }
```
Can we support writing to multiple tables if this is a comma separate list of table names? Like we do in "static" mode.

I think this is important because then we can remove regex-matcher configs as it could be implemented entirely as an SMT + dynamic mode.

bryanck replied:
I think that could be a useful feature, but I don't think it should hold up this PR. If you wanted to open an issue that would be great.

fqaiser94 commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:
```diff
@@ -0,0 +1,140 @@
+              });
+
+    } else {
+      String routeValue = extractRouteValue(record.value(), routeField);
+      if (routeValue != null) {
+        config
+            .tables()
+            .forEach(
+                tableName -> {
+                  Pattern regex = config.tableConfig(tableName).routeRegex();
+                  if (regex != null && regex.matcher(routeValue).matches()) {
+                    writerForTable(tableName, record, false).write(record);
+                  }
+                });
+      }
```
nit: is this really "static" routing? I guess it's a static list of tables, but the table/s each message is written to is determined dynamically based on the route value ...

More importantly, is this an important enough use-case to support within the connector? I would strongly prefer if we didn't support this within the connector itself (users can easily implement this by writing an SMT + dynamic mode).

bryanck replied:
Static in the sense that the list of tables is fixed and doesn't change, rather than deriving it from the record. This feature is in use by some.

bryanck commented on kafka-connect/kafka-connect-events/src/test/java/org/apache/iceberg/connect/events/EventTestUtil.java:
```diff
@@ -45,7 +45,7 @@ private EventTestUtil() {}
 
   static final PartitionSpec SPEC =
-      PartitionSpec.builderFor(SCHEMA).identity("id").withSpecId(1).build();
+      PartitionSpec.builderFor(SCHEMA).identity("id").withSpecId(0).build();
```
Thanks I changed this.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CatalogUtils.java:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  // use reflection here to avoid requiring Hadoop as a dependency
+  private static Object loadHadoopConfig(IcebergSinkConfig config) {
```
I agree, that could be useful, though it may require some discussion to get right, so I'd rather start with this here.

ajantha-bhat replied:
ok. We can move it later.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CommitterFactory.java:
```diff
@@ -0,0 +1,29 @@
+import org.apache.iceberg.connect.channel.CommitterImpl;
+
+public class CommitterFactory {
+  public static Committer createCommitter(IcebergSinkConfig config) {
```
It isn't needed now, but possibly in the future, to indicate the type of committer to create. The API is public so I designed it with that in mind.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkConfig.java:
```diff
@@ -80,7 +80,6 @@ public class IcebergSinkConfig extends AbstractConfig {
   private static final String TABLES_SCHEMA_CASE_INSENSITIVE_PROP =
       "iceberg.tables.schema-case-insensitive";
   private static final String CONTROL_TOPIC_PROP = "iceberg.control.topic";
-  private static final String CONTROL_GROUP_ID_PROP = "iceberg.control.group-id";
```
We use the connect consumer group now, rather than a separate consumer group that we keep in sync.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+  public void put(Collection<SinkRecord> sinkRecords) {
+    if (committer != null) {
+      committer.save(sinkRecords);
+    }
```
This should never happen, so I changed this to a precondition check.

fqaiser94 replied:
I can tell you that it does happen and there's nothing we can do about it other than throw an error and have the task retry. Precondition check makes sense  👍

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  @Override
+  public void flush(Map<TopicPartition, OffsetAndMetadata> currentOffsets) {
+    committer.save(null);
+  }
```
This is here in case future committer implementations want to perform an action on flush. The current committer doesn't do anything when the record collection is null.

fqaiser94 replied:
> This is here in case future committer implementations want to perform an action on flush.

What I'm trying to say is that `flush` will never be called since you're overriding the default `preCommit` implementation: https://github.com/apache/kafka/blob/c97d4ce026fdb75f80760bcce00b239db951a481/connect/api/src/main/java/org/apache/kafka/connect/sink/SinkTask.java#L139

So no committer implementation will ever be able to "perform an action on flush" unless you call `flush` yourself somewhere.

bryanck replied:
I felt it is safer to implement this, given it is a public method, rather than rely on default behavior.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+    committer.start(catalog, config, context, partitions);
+  }
+
+  @Override
+  public void close(Collection<TopicPartition> partitions) {
+    close();
+  }
+
+  private void close() {
+    if (committer != null) {
+      committer.stop();
+      committer = null;
+    }
+
+    if (catalog != null) {
+      if (catalog instanceof AutoCloseable) {
+        try {
+          ((AutoCloseable) catalog).close();
+        } catch (Exception e) {
+          LOG.warn("An error occurred closing catalog instance, ignoring...", e);
+        }
+      }
+      catalog = null;
+    }
+  }
```
I'm not sure I follow your suggestion. When the KC `close()` or `stop()` lifecycle methods are called, we close the committer and the catalog.

ajantha-bhat replied:
I think he means just rearrange the code, keep close method just before stop.

bryanck replied:
I have this order to prevent a running committer from having reference to a closed catalog.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+
+  @Override
+  public void open(Collection<TopicPartition> partitions) {
+    catalog = CatalogUtils.loadCatalog(config);
```
Thanks, I added a precondition check here to be safer.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+
+  @Override
+  public String version() {
+    return IcebergSinkConfig.version();
```
Thanks for catching this, we can just use the Iceberg version now. I left in the config method in case we want to add to this later.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SinkWriter.java:
```diff
@@ -0,0 +1,140 @@
+
+  private void routeRecordDynamically(SinkRecord record) {
+    String routeField = config.tablesRouteField();
+    Preconditions.checkNotNull(routeField, "Route field cannot be null with dynamic routing");
```
We already check this in the config, so I'll just remove this.

ajantha-bhat reviewed (approved):
LGTM. 

The required basic functionality of commit coordinator looks good to me.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/IcebergSinkTask.java:
```diff
@@ -0,0 +1,103 @@
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSinkTask.class);
+
+  private IcebergSinkConfig config;
+  private Catalog catalog;
```
A while ago there was some discussion about the thread safety of the catalog. At the time we felt we should assume it it thread safe and fix cases where it isn't. I'm not sure if that is still the case, but I'd prefer to assume that, to keep the code simpler for now. I'll raise this in Slack and we can fast follow with an update if needed.

bryanck commented:
@ajantha-bhat @fqaiser94 Thanks for the reviews! I think I have addressed comments in case you want to take another look. @Fokko @nastra If either of you have time to take a look, that would be very appreciated.

ajantha-bhat commented:
@rdblue, @danielcweeks, @Fokko, @nastra: Please take a look. It is needed for 1.6.0 milestone.

fqaiser94 reviewed (commented):
Sorry for the delay, this has been on my radar for a couple of weeks but i don't have a lot of time for reviews currently.

fqaiser94 commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:
```diff
@@ -0,0 +1,294 @@
+      LOG.info("Nothing to commit to table {}, skipping", tableIdentifier);
+    } else {
+      if (deleteFiles.isEmpty()) {
+        AppendFiles appendOp = table.newAppend();
+        if (branch != null) {
+          appendOp.toBranch(branch);
+        }
+        appendOp.set(snapshotOffsetsProp, offsetsJson);
+        appendOp.set(COMMIT_ID_SNAPSHOT_PROP, commitState.currentCommitId().toString());
+        if (vtts != null) {
+          appendOp.set(VTTS_SNAPSHOT_PROP, vtts.toString());
+        }
+        dataFiles.forEach(appendOp::appendFile);
+        appendOp.commit();
```
Alternatively, we could try getting this PR into iceberg-core: https://github.com/apache/iceberg/pull/9860
That would remove the need to do anything special here.

bryanck replied:
I feel this should be addressed in Iceberg core

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaClientFactory.java:
```diff
@@ -0,0 +1,68 @@
+    consumerProps.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+    consumerProps.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed");
+    consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupId);
+    consumerProps.put(ConsumerConfig.CLIENT_ID_CONFIG, UUID.randomUUID().toString());
```
I changed this to `putIfAbsent` so it can be overridden.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaClientFactory.java:
```diff
@@ -0,0 +1,68 @@
+  Producer<String, byte[]> createProducer(String transactionalId) {
+    Map<String, Object> producerProps = Maps.newHashMap(kafkaProps);
+    producerProps.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId);
+    producerProps.put(ProducerConfig.CLIENT_ID_CONFIG, UUID.randomUUID().toString());
```
Changed to `putIfAbsent` here also.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Worker.java:
```diff
@@ -0,0 +1,127 @@
+    events.add(readyEvent);
+
+    send(events, results.sourceOffsets());
+    context.requestCommit();
```
I removed this, it is no longer necessary.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java:
```diff
@@ -0,0 +1,294 @@
+    this.snapshotOffsetsProp =
+        String.format(
+            "kafka.connect.offsets.%s.%s", config.controlTopic(), config.connectGroupId());
+    this.exec = ThreadPools.newWorkerPool("iceberg-committer", config.commitThreads());
```
This uses daemon threads.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    if (System.currentTimeMillis() - startTime > config.commitTimeoutMs()) {
+      LOG.info("Commit timeout reached");
```
Added

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/KafkaUtils.java:
```diff
@@ -0,0 +1,61 @@
+              DynFields.builder().hiddenImpl(CONTEXT_CLASS_NAME, "consumer").build(context).get())
+          .groupMetadata();
+    }
+    return new ConsumerGroupMetadata(connectGroupId);
```
I added a warn log statement.

bryanck replied:
I removed the fallback, so it hard fails now.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:
```diff
@@ -0,0 +1,171 @@
+        recordList.forEach(producer::send);
+        if (!sourceOffsets.isEmpty()) {
+          producer.sendOffsetsToTransaction(
+              offsetsToCommit, KafkaUtils.consumerGroupMetadata(context, connectGroupId));
```
We need to update the metadata to get the latest generation ID.

danielcweeks commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:
```diff
@@ -0,0 +1,171 @@
+    consumer.commitSync(offsetsToCommit);
+  }
+
+  protected Admin admin() {
```
No usages for this, is it necessary?

bryanck replied:
I removed this, it isn't needed anymore.

danielcweeks commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java:
```diff
@@ -0,0 +1,167 @@
+              .flatMap(event -> event.assignments().stream())
+              .map(TopicPartitionOffset::timestamp)
+              .min(Comparator.naturalOrder())
+              .get();
```
Are we sure this `get()` is safe?  The `allMatch` above will return ture if there are no elements.

bryanck replied:
I changed this to `.orElse(null)`

danielcweeks reviewed (commented):
Some minor nits and questions, but overall LGTM.

I think there are still a couple open questions from others (particularly the one on `IcebergSinkTask.open()`, which we probably want to address.

I did see that we renamed a few classes and reduced the visibility in others.  I'm ok with that for now since this isn't really deployable, but we should probably avoid that going forward.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/CommitterFactory.java:
```diff
@@ -0,0 +1,29 @@
+import org.apache.iceberg.connect.channel.CommitterImpl;
+
+public class CommitterFactory {
+  public static Committer createCommitter(IcebergSinkConfig config) {
```
You're right, I reduced visibility to package private.

bryanck commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/CommitState.java:
```diff
@@ -0,0 +1,167 @@
+                envelope -> ((DataWritten) envelope.event().payload()).tableReference()));
+  }
+
+  OffsetDateTime vtts(boolean partialCommit) {
```
I replaced all occurrences of `vtts` .

bryanck commented:
I think I've addressed the latest feedback @fqaiser94 @danielcweeks LMK if there is anything else.

eshishki commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Channel.java:
```diff
@@ -0,0 +1,167 @@
+          record -> {
+            // the consumer stores the offsets that corresponds to the next record to consume,
+            // so increment the record offset by one
+            controlTopicOffsets.put(record.partition(), record.offset() + 1);
```
why do actually need to increment? shouldn't we instead decrement to get current offset
